### ISSUE TITLE
json encoding differences cause canary value test to fail

### ISSUE THREAD

**Issue by PaulTurner-awin** [ticarpi/jwt_tool#48]
Firstly, thanks for a great tool!

I'm not sure there's much that can be done about this, but I wanted to log it in case anyone else has the same experience.

If the original JWT token includes a (correctly) escaped forward slash in the issuer, such as ` {"iss":"my.domain.com\/iss-endpoint"}`, then jwt_tool fails to recognise the canary value test has been met, such as:

> [+] Sending token
> jwttool_b5c2514b08df054d76b80e7d6a98c70d Sending token Response Code: 403, 91 bytes
> Running Scanning Module:
> Running prescan checks...
> [+] FOUND "clickref" in response:
> jwttool_86deb84549be4dcfd8661a928c5b76f8 Prescan: original token Response Code: 200, 254 bytes
> Canary value (myvalue) was not found in base request - check that this token is valid and you are still logged in
> Do you wish to continue anyway? ("Y" or "N")
Here you can see the original request failed with a 403 response, this seems to be due to how the[ rejigToken](https://github.com/ticarpi/jwt_tool/blob/4045e9c6df5cf079c3ea6cd30e2c774c85a7f0e2/jwt_tool.py#L2028) function works.

Having spent sometime looking in to this, it appears to be due to the json decoding and re-encoding not resulting in the same result, which invalidates the original signature.

This Python code may help illustrate it more:
```
import json

json_string = '{"iss":"my.domain.com\/iss-endpoint"}'#

json_object = json.loads(json_string)

new_json_string = json.dumps(json_object)

if json_string == new_json_string:
        print("json remains unchanged:" + new_json_string)
else:
        print("json had changed!  original:"+ json_string + "  new:" + new_json_string)
```
Running this will output:

> $ python example.py
> json had changed!  original:{"iss":"my.domain.com\\/iss-endpoint"}  new:{"iss": "my.domain.com/iss-endpoint"}

For clarity my jwt_tool command was like:
`python jwt_tool.py -pd "{MY POST DATA}" -t {MY URL}  -rh "Authorization: Bearer {MY JWT}" -M pb -cv "myvalue"`

The escaping of the forward slash as "\/" is valid (but optional) according to [the specification](https://www.ecma-international.org/wp-content/uploads/ECMA-404_2nd_edition_december_2017.pdf).

I'm not sure whether it would be possible to "fix" this in jwt_tool or not?

**Comment by dnet** [comment]
I have ran into this same issue, a KeyCloak instance has used a different JSON serializer than the one JWT_Tool uses:

```
% hexyl jwt_tool.dec
┌────────┬─────────────────────────┬─────────────────────────┬────────┬────────┐
│00000000│ 7b 22 61 6c 67 22 3a 22 ┊ 52 53 32 35 36 22 2c 22 │{"alg":"┊RS256","│
│00000010│ 74 79 70 22 3a 22 4a 57 ┊ 54 22 2c 22 6b 69 64 22 │typ":"JW┊T","kid"│
│00000020│ 3a 22 6c 4b 4a 33 6a 30 ┊ 64 4f 4b 53 67 70 79 7a │:"lKJ3j0┊dOKSgpyz│
│00000030│ 33 74 41 30 57 66 6c 6b ┊ 68 56 75 4f 52 51 67 43 │3tA0Wflk┊hVuORQgC│
│00000040│ 75 39 78 32 37 6c 47 62 ┊ 77 74 43 41 34 22       │u9x27lGb┊wtCA4"  │
└────────┴─────────────────────────┴─────────────────────────┴────────┴────────┘
% hexyl burp.dec
┌────────┬─────────────────────────┬─────────────────────────┬────────┬────────┐
│00000000│ 7b 22 61 6c 67 22 3a 22 ┊ 52 53 32 35 36 22 2c 22 │{"alg":"┊RS256","│
│00000010│ 74 79 70 22 20 3a 20 22 ┊ 4a 57 54 22 2c 22 6b 69 │typ" : "┊JWT","ki│
│00000020│ 64 22 20 3a 20 22 6c 4b ┊ 4a 33 6a 30 64 4f 4b 53 │d" : "lK┊J3j0dOKS│
│00000030│ 67 70 79 7a 33 74 41 30 ┊ 57 66 6c 6b 68 56 75 4f │gpyz3tA0┊WflkhVuO│
│00000040│ 52 51 67 43 75 39 78 32 ┊ 37 6c 47 62 77 74 43 41 │RQgCu9x2┊7lGbwtCA│
│00000050│ 34                      ┊                         │4       ┊        │
└────────┴─────────────────────────┴─────────────────────────┴────────┴────────┘
```

As it can be seen, spaces (ASCII `0x20`) are inserted around the colons after both `typ` and `kid`, while JWT_Tool serializes the JWT header using the most compact format that uses no whitespace within the JSON structure. While latter matches the behavior of most common JWT libraries, [RFC 7519](https://tools.ietf.org/html/rfc7519) (which describes JWT) mentions no such rule, au contraire:

> This JSON object MAY contain whitespace and/or line breaks before or after any JSON values or structural characters, in accordance with [Section 2 of RFC 7159](https://datatracker.ietf.org/doc/html/rfc7159#section-2)

> Note that whitespace is explicitly allowed in the representation and no canonicalization need be performed before encoding.

However, JWT_Tool does exactly that, performing a canonicalization (c14n) by (re)serializing JSON data even in case of the canary value test. Since the hash needs to be calculated on the base64-encoded JSON output, this inevitably breaks the hash, resulting in `401 Unauthorized` responses.

How should this be solved? Tough call, maybe the canary check should avoid the deserialization-(re)serialization loop altogether. Another option is to parse in a way that keeps whitespace (meta)information and (re)applies it during serialization, like XML libraries do.

**Comment by dnet** [comment]
> The problem with the whitespaces in your example is mainly, that they are not consistent.

I agree, this was the most baffling thing for me as well. Then again, I did not write the implementation, I just want to test it – and optionally break it.

In the end, I don't really understand the need for the aforementioned deserialization-(re)serialization loop. That's just unnecessary processing that goes directly against what the standard says and breaks edge cases like this.
